pkg/service/presentation/service.go

```diff
@@ -104,25 +127,72 @@ func (s Service) DeletePresentationDefinition(request DeletePresentationDefiniti
 
 // CreateSubmission houses the main service logic for presentation submission creation. It validates the input, and
 // produces a presentation submission value that conforms with the Submission specification.
-func (s Service) CreateSubmission(request CreateSubmissionRequest) (*CreateSubmissionResponse, error) {
-	logrus.Debugf("creating presentation submission: %+v", request)
-
+func (s Service) CreateSubmission(request CreateSubmissionRequest) (*operation.Operation, error) {
 	if !request.IsValid() {
-		return nil, util.LoggingNewErrorf("invalid create presentation submission request: %+v", request)
+		return nil, errors.Errorf("invalid create presentation submission request: %+v", request)
 	}
 
 	if err := exchange.IsValidPresentationSubmission(request.Submission); err != nil {
-		return nil, util.LoggingErrorMsg(err, "provided value is not a valid presentation submission")
+		return nil, errors.Wrap(err, "provided value is not a valid presentation submission")
+	}
+
+	sdkVP, err := signing.ParseVerifiablePresentationFromJWT(request.SubmissionJWT.String())
+	if err != nil {
+		return nil, errors.Wrap(err, "parsing vp from jwt")
+	}
+	if err := jwt.VerifyTokenFromDID(sdkVP.Holder, request.SubmissionJWT, s.resolver); err != nil {
+		return nil, errors.Wrap(err, "verifying token from did")
+	}
+
+	if _, err := s.storage.GetSubmission(request.Submission.ID); !errors.Is(err, presentationstorage.ErrSubmissionNotFound) {
+		return nil, errors.Errorf("submission with id %s already present", request.Submission.ID)
+	}
+
+	definition, err := s.storage.GetPresentation(request.Submission.DefinitionID)
+	if err != nil {
+		return nil, errors.Wrap(err, "getting presentation definition")
+	}
+
+	for _, cred := range request.Credentials {
+		if !cred.IsValid() {
+			return nil, errors.Errorf("invalid credential %+v", cred)
+		}
+		if cred.CredentialJWT != nil {
+			if err := s.verifier.VerifyJWTCredential(*cred.CredentialJWT); err != nil {
+				return nil, errors.Wrapf(err, "verifying jwt credential %s", cred.CredentialJWT)
+			}
+		} else {
```

Comment:
can get rid of this else here and have the same behavior

Reply:
Thanks for the suggestion! I'm keeping since `cred` might have both `CredentialJWT`  and `Credential` populated, so having the else would avoid doing double validation in that case.